src/quicklook.cs

```diff
@@ -170,6 +171,87 @@ interface QLPreviewItem {
 #endif
 	}
 
+	delegate bool QLPreviewReplyDrawingHandler (CGContext context, QLPreviewReply reply, out NSError error);
+	delegate NSData QLPreviewReplyDataCreationHandler (QLPreviewReply reply, out NSError error);
+	delegate CGPDFDocument QLPreviewReply_UIDocumentCreationHandler (QLPreviewReply reply, out NSError error);
+
+	[NoWatch, NoTV, Mac (12,0), iOS (15,0), MacCatalyst (15,0)]
+	[BaseType (typeof(NSObject))]
+	interface QLPreviewReply
+	{
+		[Export ("stringEncoding")]
+		nuint StringEncoding { get; set; }
+
+		[Export ("attachments", ArgumentSemantic.Copy)]
+		NSDictionary<NSString, QLPreviewReplyAttachment> Attachments { get; set; }
+
+		[Export ("title")]
+		string Title { get; set; }
+
+		[Export ("initWithContextSize:isBitmap:drawingBlock:")]
+		IntPtr Constructor (CGSize contextSize, bool isBitmap, QLPreviewReplyDrawingHandler drawingHandler);
+
+		[Export ("initWithFileURL:")]
+		IntPtr Constructor (NSUrl fileURL);
+
+		[Export ("initWithDataOfContentType:contentSize:dataCreationBlock:")]
+		IntPtr Constructor (UTType contentType, CGSize contentSize, QLPreviewReplyDataCreationHandler dataCreationHandler);
+
+		// QLPreviewReply_UI
+		[Export ("initForPDFWithPageSize:documentCreationBlock:")]
+		IntPtr Constructor (CGSize defaultPageSize, QLPreviewReply_UIDocumentCreationHandler documentCreationHandler);
+	}
+
+	[NoWatch, NoTV, Mac (12,0), iOS (15,0), MacCatalyst (15,0)]
+	[BaseType (typeof(NSObject))]
+	interface QLPreviewReplyAttachment
+	{
+		[Export ("data")]
+		NSData Data { get; }
+
+		[Export ("contentType")]
+		UTType ContentType { get; }
+
+		[Export ("initWithData:contentType:")]
+		IntPtr Constructor (NSData data, UTType contentType);
+	}
+
+	[NoWatch, NoTV, Mac (12,0), iOS (15,0), MacCatalyst (15,0)]
+	[BaseType (typeof(NSObject))]
+	interface QLFilePreviewRequest
+	{
+		[Export ("fileURL")]
+		NSUrl FileURL { get; }
+	}
+
+	[NoWatch, NoTV, Mac (12,0), iOS (15,0), MacCatalyst (15,0)]
+	[BaseType (typeof(NSObject))]
+	interface QLPreviewProvider : NSExtensionRequestHandling
+	{
+	}
+
+	[iOS (15,0), MacCatalyst (15,0)]
```

Comment:
Mac ?

without a `[Mac]` it means the oldest supported (10,9) has it available.

Same for watchOS and tvOS. You likely want `[NoWatch, NoTV, NoMac]` if it's not available on macOS.

Reply:
if there's no indication on the diffs, xtro, web docs, or header files - I tend to leave the availability of unmentioned platforms as is. But is it best to assume that there is no availability on those platforms? Since it's a breaking change to remove things, and it's always easier to introduce things later on?

Reply:
> But is it best to assume that there is no availability on those platforms? 

No. Lack of availability attributes means it's been available since our earliest supported versions.
E.g. lack of `[NoWatch]` implies it's `[Watch (2,0)]` and that the API has been available in the past 6 years (or so)

This is almost always wrong when adding bindings for a new Xcode (unless it added a new platform).

> Since it's a breaking change to remove things, and it's always easier to introduce things later on?

True :-) However it's true for code that we ship (post generation), not the binding sources!

Example:  `[NoWatch]` -> means we're not generating the API on watchOS.

If you don't _add_ `[NoWatch]` (in binding sources) then the API will be present inside `Xamarin.WatchOS.dll` (and assumed to be available since 2.0 which is watchOS minimum supported version).

If you want to fix that later, by adding `[NoWatch]`, then you are creating a **breaking change** (since the API won't be part of `Xamarin.WatchOS.dll` in the next release).

Reply:
Ah, I think my question was a bit confusing! I understand what you explained about the lack of availability attributes in our code and I meant to ask about how to interpret the lack of mention of platform availability in diffs, xtro, web docs, and header files. 

So in this case, there is no Watch availability or unavailability mentioned whatsoever in the diffs, xtro, web docs, or header files for this API. So I was asking if that means it's best to assume there's no watch availability and tag as [NoWatch] - does that make sense?

Reply:
Short answer: Whenever in doubt it's better to assume `[No*]` if availability is not mentioned because it means no code will be generated (and it won't be a breaking change to add it later).

When you add (or modify) sharpie's output then it's good to add comments (for reviewers and people who will update the bindings... years later), e.g.

```csharp
[NoWatch][NoTV][NoMac] // availability not mentioned in the header files
```


Long answer

* Apple's macro are often incomplete or incorrect (e.g. wrong version)
  * Some (not all) gets fixed in newer betas or the next Xcode...
* There are assumptions like
  * old (pre 9.0) iOS API are assumed to be available on tvOS 9 (first version) unless decorated otherwise
  * old (pre 13) iOS API are assumed to be available on MacCatalyst 13 (first version) ...
  * enums are not always decorated since they are not API (you can see which API uses them to know which attributes are needed)
* Intro (runtime) and xtro (build time) will spot many, but not all, errors
* over time you get to know the API/framework and can identify things that looks unusual (and add tests to confirm your gut feelings)

Reply:
perfect, thank you!!!